source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -49,25 +57,48 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
   lookup_ = cache_.makeLookupContext(std::move(lookup_request));
 
   ASSERT(lookup_);
-
   ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders starting lookup", *decoder_callbacks_);
-  lookup_->getHeaders([this](LookupResult&& result) { onHeaders(std::move(result)); });
-  if (state_ == GetHeadersState::GetHeadersResultUnusable) {
-    // onHeaders has already been called, and no usable cache entry was found--continue iteration.
-    return Http::FilterHeadersStatus::Continue;
+
+  lookup_->getHeaders(
+      [this, &headers](LookupResult&& result) { onHeaders(std::move(result), headers); });
+
+  if (filter_state_ == FilterState::Initial) {
```

Comment:
It seems surprising to check filter_state_ here. I suggest explaining it with a comment like // If the plugin called onHeaders synchronously, onHeaders will have advanced filter_state_.